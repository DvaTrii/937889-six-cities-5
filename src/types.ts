export type OfferCardType = {
  id: number,
  city: {
    name: string,
    coordinates: {
      latitude: number,
      longitude: number,
      zoom: number
    }
  },
  coordinates: {
    latitude: number,
    longitude: number,
    zoom: number
  },
  pictures: Array<string>,
  previewImage: string,
  isPremium?: boolean,
  isBookmark?: boolean,
  price: number,
  title: string,
  type: string,
  rating: number,
  description: string,
  bedroomsMax: number,
  guestsMax: number,
  amenities: Array<string> | null,
  hostInfo: {
    avatar: string,
    name: string,
    isSuper?: boolean,
  }
};

export enum Cities {
  PARIS = `Paris`,
  COLOGNE = `Cologne`,
  BRUSSELS = `Brussels`,
  AMSTERDAM = `Amsterdam`,
  HAMBURG = `Hamburg`,
  DUSSELDORF = `Dusseldorf`,
}

export type ReviewType = {
  id: number,
  date: number,
  text: string,
  rating: number,
  userInfo: {
    avatar: string,
    name: string,
    id: number,
    isSuper: boolean,
  }
};
