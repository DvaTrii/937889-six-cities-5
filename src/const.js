export const OfferType = {
  APARTMENT: `Apartment`,
  ROOM: `Room`,
  HOUSE: `House`,
  HOTEL: `Hotel`
};
