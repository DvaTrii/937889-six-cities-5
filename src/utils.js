export const extend = (a, b) => {
  return Object.assign({}, a, b);
};
