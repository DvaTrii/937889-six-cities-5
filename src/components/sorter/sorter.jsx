import React from "react";
import {connect} from "react-redux";
import {compose} from "redux";
import PropTypes from "prop-types";

import {withActiveFlag} from "../hocs/withActiveFlag/withActiveFlag";
import {SorterType, SorterActiveClass, SorterListOpenedClass} from "../../const";
import {setActiveSorter} from "../../store/app/actions";


const Sorter = ({isActive, onActiveChange, activeSorter, setActiveSorterAction}) => {
  return (
    <form className="places__sorting" action="#" method="get">
      <span className="places__sorting-caption">Sort by</span>
      <span className="places__sorting-type"
        tabIndex="0"
        onClick={onActiveChange}>
        {activeSorter}
        <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"></use>
        </svg>
      </span>
      <ul className={`places__options places__options--custom
        ${isActive ?
      SorterListOpenedClass.OPENED : ``}`}>

        {
          Object.values(SorterType).map((it) => (
            <li key={it}
              className={`places__option ${activeSorter === it ? SorterActiveClass.ACTIVE : ``}`}
              onClick={() => {
                setActiveSorterAction(it);
                onActiveChange();
              }}
              tabIndex="0">
              {it}
            </li>))
        }

      </ul>
      {/* <select className="places__sorting-type" id="places-sorting">*/}
      {/*  <option className="places__option" value="popular" selected="">Popular</option>*/}
      {/*  <option className="places__option" value="to-high">Price: low to high</option>*/}
      {/*  <option className="places__option" value="to-low">Price: high to low</option>*/}
      {/*  <option className="places__option" value="top-rated">Top rated first</option>*/}
      {/* </select>*/}
    </form>
  );
};

Sorter.propTypes = {
  isActive: PropTypes.bool.isRequired,
  onActiveChange: PropTypes.func.isRequired,
  activeSorter: PropTypes.string.isRequired,
  setActiveSorterAction: PropTypes.func.isRequired
};

const mapStateToProps = ({APP}) => {
  return ({
    activeSorter: APP.activeSorter,
  });
};

const mapDispatchToProps = (dispatch) => ({
  setActiveSorterAction(sorter) {
    dispatch(setActiveSorter(sorter));
  }
});

export {Sorter};
export default compose(withActiveFlag, connect(mapStateToProps, mapDispatchToProps))(Sorter);
