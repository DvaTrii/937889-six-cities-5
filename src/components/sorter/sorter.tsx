import * as React from "react";
import {connect} from "react-redux";

import {SorterType, SorterActiveClass, SorterListOpenedClass} from "../../const";
import {getActiveSorter, getActiveCity} from "../../store/app/selectors";
import {setActiveSorter} from "../../store/app/actions";

type Props = {
  activeSorter: string,
  setActiveSorterAction: (string) => void,
  activeCity: string
}

const Sorter: React.FC<Props> = ({activeSorter, setActiveSorterAction, activeCity}) => {

  const [isActive, setIsActive] = React.useState<boolean>(false);
  const toggleIsActive = () => setIsActive(!isActive);

  React.useEffect(() => {
    setIsActive(false);
  }, [activeCity]);

  return (
    <form className="places__sorting" action="#" method="get">
      <span className="places__sorting-caption">Sort by</span>
      <span className="places__sorting-type"
        // tabIndex="0"
        onClick={toggleIsActive}>
        {activeSorter}
        <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"/>
        </svg>
      </span>
      <ul className={`places__options places__options--custom
        ${isActive ?
      SorterListOpenedClass.OPENED : ``}`}>

        {
          Object.values(SorterType).map((it: string) => (
            <li key={it}
              className={`places__option ${activeSorter === it ? SorterActiveClass.ACTIVE : ``}`}
              onClick={() => {
                setActiveSorterAction(it);
                toggleIsActive();
              }}
              // tabIndex="0"
              >
              {it}
            </li>))
        }

      </ul>
    </form>
  );
};

const mapStateToProps = (state) => {
  return ({
    activeSorter: getActiveSorter(state),
    activeCity: getActiveCity(state)
  });
};

const mapDispatchToProps = (dispatch) => ({
  setActiveSorterAction(sorter) {
    dispatch(setActiveSorter(sorter));
  }
});

export {Sorter};
export default connect(mapStateToProps, mapDispatchToProps)(Sorter);
