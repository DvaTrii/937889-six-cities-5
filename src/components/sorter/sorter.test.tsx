import * as React from "react";
import * as renderer from "react-test-renderer";
import {Sorter} from "./sorter";
import {noop} from "../../utils";
import {Cities} from "../../types";
import {SorterType} from "../../const";

it(`Should Sorter render properly`, () => {

  const activeSorter: string = SorterType.Popular;
  const activeCity: Cities = Cities.PARIS;

  const tree = renderer
    .create(
        <Sorter
          activeSorter={activeSorter}
          setActiveSorterAction={noop}
          activeCity={activeCity}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
