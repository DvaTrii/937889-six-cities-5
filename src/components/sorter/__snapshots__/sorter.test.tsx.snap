// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Sorter render properly 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
>
  <span
    className="places__sorting-caption"
  >
    Sort by
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
  >
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom
        "
  >
    <li
      className="places__option "
      onClick={[Function]}
    >
      Popular
    </li>
    <li
      className="places__option "
      onClick={[Function]}
    >
      Price: low to high
    </li>
    <li
      className="places__option "
      onClick={[Function]}
    >
      Price: high to low
    </li>
    <li
      className="places__option "
      onClick={[Function]}
    >
      Top rated first
    </li>
  </ul>
</form>
`;
